rfcs/20200721-extension-types.md

```diff
@@ -0,0 +1,1304 @@
+`expand_composites`.  This argument indicates that composite tensors should be
+treated as nested structures, and expanded into their component `Tensors`; and
+similarly, that `TypeSpecs` should be treated as nested structures, and expanded
+into their component `TensorSpecs`.  In particular:
+
+**tf.nest.flatten:**
+
+* `tf.nest.flatten(composite_tensor, expand_composites=True)` returns a flat
+  list of the `tf.Tensor` components from `composite_tensor`.
+* `tf.nest.flatten(type_spec, expand_composites=True)` returns a flat list of
+  `tf.TensorSpecs` describing the tensor components for `type_spec`.
+
+**tf.nest.pack_sequence_as:**
+
+* `tf.nest.pack_sequence_as(type_spec, tensor_list, expand_composites=True)`
+  uses `type_spec.from_components` to reconstruct a composite tensor from its
+  components.  Note that the new value's dynamic (tensor) data will come from
+  `tensor_list`, but static (non-tensor) data will come from `type_spec`.
+
+* `tf.nest.pack_sequence_as(composite_tensor, tensor_list, expand_composites=True)`
+  uses `composite_tensor.__tf_type_spec__().from_components` to reconstruct the
+  CompositeTensor from components.
+
+**Note:** When using `tf.nest.pack_sequence_as` with composite tensors, the
+`flat_sequence` argument must be a list of `Tensor`; it may not be a list of
+`TensorSpec`.
+
+**tf.nest.assert_same_structure:**
+
+* If `x` and `y` are both composite tensors or `TypeSpecs`, then
+  `tf.nest.assert_same_structure(x, y, expand_composites=True)` raises an
+  exception if there is no `TypeSpec` compatible with both `x` and `y` (as
+  determined by calling `TypeSpec.most_specific_compatible_type`).
+
+**tf.nest.map_structure:**
+
+* `tf.nest.map_structure(func, composite_tensor, expand_composites=True)`
+  transforms `composite_tensor` by flattening it into its component tensors,
+  applying `func` to transform each component tensor, and then repacking those
+  transformed tensors into a composite tensor with the original type.
+
+The following example uses `nest.flatten` with `expand_composites=True` to
+convert a nested structure containing composite tensors to a list of
+`tf.Tensors`; applies a function `f` to transform each tensor; and then uses
+`nest.pack_sequence_as` with `expand_composites=True` to reassemble the results
+back into the original structure.
+
+```python
+>>> rt = RaggedTensor(values=v1, row_splits=r)
+>>> st = SparseTensor(indices=i, values=v2, dense_shape=d)
+>>> structure = {'a': rt, 'b': st}
+>>> flat = nest.flatten(structure, expand_composites=True)
+[v1, r, i, v2, d]
+>>> mapped = [f(t) for t in flat]
+>>> nest.pack_sequence_as(structure, mapped)
+{'a': RaggedTensor(f(v1), f(r)), 'b': SparseTensor(f(i), f(v2), f(d))}
+```
+
+#### TypeSpec Registry
+
+In order to be used with `SavedModels`, extension types must register their
+`TypeSpecs` using `tf.register_type_spec`.
+
+```python
+def register_type_spec(type_spec_subclass, name=None):
```

Comment:
This would be neat to have as a class decorator. It would link the otherwise decoupled declaration and further registration of the class. (Also could be used to build a centralized list of all declarations at compile/build time in a relatively ease way, but not sure about that).

Reply:
Agreed -- I think this should be a decorator.  (Since the common case is that `name` is None, the only thing we really need to do to make it a class decorator is have it return `type_spec_subclass` rather than `None`.)